File: src/main/python/net/x3djsonld/data/c/x3d-code/www.web3d.org/x3d/content/examples/Basic/development/MaterialExample.py
###############################################
#
# Now available: developmental python x3d.py package on PyPi for import.
#   This approach greatly simplifies Python X3D deployment and use.
#   https://pypi.org/project/x3d
#
# Installation:
#       pip install x3d
# or
#       python -m pip install x3d
#
# Developer options for loading x3d package:
#
#    from x3d import *  # preferred approach, terser source that avoids x3d.* class prefixes
#
# or
#    import x3d         # traditional way to subclass x3d package, all classes require x3d.* prefix
#                       # but python source is very verbose, for example x3d.Material x3d.Shape etc.
#                       # X3dToPython.xslt stylesheet insertPackagePrefix=true supports this option.

from x3d import *

###############################################

newModel=X3D(profile='Immersive',version='3.2',
  head=head(
    children=[
    meta(content='MaterialExample.x3d',name='title'),
    meta(content='Material node example for X3D-Edit MaterialEditor, which uses Scene Access Interface (SAI) externally to select/deselect different shapes, axes and lights while modifying Material values.',name='description'),
    meta(content='Don Brutzman',name='creator'),
    meta(content='17 December 2006',name='created'),
    meta(content='20 October 2019',name='modified'),
    meta(content='X3D-Edit example',name='subject'),
    meta(content='https://www.web3d.org/x3d/content/examples/Basic/development/MaterialExample.x3d',name='identifier'),
    meta(content='src/www.web3d.org/x3d/tools/X3dEdit3.3/X3D/src/org/web3d/x3d/templates/MaterialExample.x3d',name='reference'),
    meta(content='X3D-Edit 3.3, https://savage.nps.edu/X3D-Edit',name='generator'),
    meta(content='../license.html',name='license')]),
  Scene=Scene(
    children=[
    WorldInfo(title='MaterialExample.x3d'),
    Viewpoint(DEF='ViewpointNode',description='default',position=(0,0,4)),
    NavigationInfo(headlight=False,type=["EXAMINE"]),
    Background(DEF='BackgroundNode'),
    DirectionalLight(DEF='DirectionalLightNode',direction=(-0.707,0,-0.707)),
    Shape(DEF='ShapeNode',
      geometry=Sphere(),
      appearance=Appearance(
        material=Material(DEF='MaterialNode'))),
    Switch(DEF='AxesSwitch',whichChoice=0,
      children=[
      Group(
        children=[
        Group(DEF='xAxis',
          children=[
          Shape(
            geometry=IndexedLineSet(coordIndex=[0,1],
              coord=Coordinate(point=[(0,0,0),(1.25,0,0)])),
            appearance=Appearance(DEF='AppearanceRed',
              material=Material(diffuseColor=(0,0,0),emissiveColor=(0.8,0,0)))),
          Transform(translation=(1.4,0,0),
            children=[
            Billboard(
              children=[
              Shape(
                geometry=Text(string=["x"],
                  fontStyle=FontStyle(DEF='LabelStyle',justify=["MIDDLE","MIDDLE"],size=0.3)),
                appearance=Appearance(USE='AppearanceRed'))])])]),
        Group(DEF='yAxis',
          children=[
          Shape(
            geometry=IndexedLineSet(coordIndex=[0,1],
              coord=Coordinate(point=[(0,0,0),(0,1.25,0)])),
            appearance=Appearance(DEF='AppearanceGreen',
              material=Material(diffuseColor=(0,0,0),emissiveColor=(0,0.8,0)))),
          Transform(translation=(0,1.4,0),
            children=[
            Billboard(axisOfRotation=(0,0,0),
              children=[
              Shape(
                geometry=Text(string=["y"],
                  fontStyle=FontStyle(USE='LabelStyle')),
                appearance=Appearance(USE='AppearanceGreen'))])])]),
        Group(DEF='zAxis',
          children=[
          Shape(
            geometry=IndexedLineSet(coordIndex=[0,1],
              coord=Coordinate(point=[(0,0,0),(0,0,1.25)])),
            appearance=Appearance(DEF='AppearanceBlue',
              material=Material(diffuseColor=(0,0,0),emissiveColor=(0,0,0.8)))),
          Transform(translation=(0,0,1.4),
            children=[
            Billboard(axisOfRotation=(0,0,0),
              children=[
              Shape(
                geometry=Text(string=["z"],
                  fontStyle=FontStyle(USE='LabelStyle')),
                appearance=Appearance(USE='AppearanceBlue'))])])])])]),
    Switch(DEF='LightDirectionVectorSwitch',whichChoice=0,
      children=[
      Transform(scale=(1.25,1.25,1.25),
        children=[
        Shape(
          geometry=IndexedLineSet(coordIndex=[0,1],
            #  negate direction-vector values so that line approaches from light direction towards origin 
            coord=Coordinate(DEF='LightDirectionCoordinateNode',point=[(0.707,0,0.707),(0,0,0)])),
          appearance=Appearance(
            material=Material(DEF='LightDirectionLineMaterial',diffuseColor=(0,0,0),emissiveColor=(1,1,1))))])])])
) # X3D model complete

###############################################
# Self-test diagnostics
###############################################

if        metaDiagnostics(newModel): # built-in utility method in X3D class
    print(metaDiagnostics(newModel))
print('check  newModel.XML() serialization...')
newModelXML = newModel.XML() # test export method XML() for exceptions
# print(newModelXML) # debug

print ("python x3d.py load successful for MaterialExample.py")

File: src/main/node/net/x3djsonld/data/c/x3d-code/www.web3d.org/x3d/content/examples/Basic/development/MaterialExample.js
var java = require('java');
java.asyncOptions = {
  asyncSuffix: undefined,     // Don't generate node-style methods taking callbacks
  syncSuffix: "",              // Sync methods use the base name(!!)
  promiseSuffix: "Promise",   // Generate methods returning promises, using the suffix Promise.
  promisify: require('util').promisify, // Needs Node.js version 8 or greater, see comment below
  ifReadOnlySuffix: "_alt"
};
var autoclass = require('./X3Dautoclass');

// Javadoc annotations follow, see below for source.
/**
 * <p> Material node example for X3D-Edit MaterialEditor, which uses Scene Access Interface (SAI) externally to select/deselect different shapes, axes and lights while modifying Material values. </p>
 <p> Related links: <a href="../../../development/MaterialExample.java">MaterialExample.java</a> source, <a href="../../../development/MaterialExampleIndex.html" target="_top">MaterialExample catalog page</a>, <a href="https://www.web3d.org/x3d/content/examples/X3dResources.html" target="_blank">X3D Resources</a>, <a href="https://www.web3d.org/x3d/content/examples/X3dSceneAuthoringHints.html" target="_blank">X3D Scene Authoring Hints</a> and <a href="https://www.web3d.org/x3d/content/X3dTooltips.html" target="_blank">X3D Tooltips</a>. </p>
	<table style="color:black; border:0px solid; border-spacing:10px 0px;" summary="Scene Metadata">
		<tr style="background-color:silver; border-color:silver;">
			<td style="text-align:center; padding:10px 0px;"><i>meta tags</i></td>
			<td style="text-align:left;   padding:10px 0px;">&nbsp; Document Metadata </td>
		</tr>

		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> title </i> </td>
			<td> <a href="../../../development/MaterialExample.x3d">MaterialExample.x3d</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> description </i> </td>
			<td> Material node example for X3D-Edit MaterialEditor, which uses Scene Access Interface (SAI) externally to select/deselect different shapes, axes and lights while modifying Material values. </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> creator </i> </td>
			<td> Don Brutzman </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> created </i> </td>
			<td> 17 December 2006 </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> modified </i> </td>
			<td> 20 October 2019 </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> subject </i> </td>
			<td> X3D-Edit example </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> identifier </i> </td>
			<td> <a href="https://www.web3d.org/x3d/content/examples/Basic/development/MaterialExample.x3d" target="_blank">https://www.web3d.org/x3d/content/examples/Basic/development/MaterialExample.x3d</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> reference </i> </td>
			<td> <a href="../../../development/src/www.web3d.org/x3d/tools/X3dEdit3.3/X3D/src/org/web3d/x3d/templates/MaterialExample.x3d">src/www.web3d.org/x3d/tools/X3dEdit3.3/X3D/src/org/web3d/x3d/templates/MaterialExample.x3d</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> generator </i> </td>
			<td> X3D-Edit 3.3, <a href="https://savage.nps.edu/X3D-Edit" target="_blank">https://savage.nps.edu/X3D-Edit</a> </td>
		</tr>
		<tr>
			<td style="text-align:right; vertical-align: text-top;"> <i> license </i> </td>
			<td> <a href="../../../development/../license.html">../license.html</a> </td>
		</tr>
		<tr style="background-color:silver; border-color:silver;">
			<td style="text-align:center;" colspan="2">  &nbsp; </td>
		</tr>
	</table>

	<p>
		This program uses the
		<a href="https://www.web3d.org/specifications/java/X3DJSAIL.html" target="_blank">X3D Java Scene Access Interface Library (X3DJSAIL)</a>.
		It has been produced using the 
		<a href="https://www.web3d.org/x3d/stylesheets/X3dToNodeJS.xslt" target="_blank">X3dToNodeJS.xslt</a>
		stylesheet to create NodeJS source code from an <code>.x3d</code> scene.
	</p>

	* @author Don Brutzman
 */

function MaterialExample
  /** Default constructor to create this object. */
  ()
  {
    /** The initialized model object, created within initialize() method. */
    this.x3dModel = {};
    this.initialize();
    return this;
  }
MaterialExample.prototype = {
  /** Create and initialize the X3D model. */
  initialize : function ()
  {
  this.x3dModel = (new autoclass.X3D()).setProfile("Immersive").setVersion("3.2")
  .setHead((new autoclass.head())
    .addMeta((new autoclass.meta()).setName("title").setContent("MaterialExample.x3d"))
    .addMeta((new autoclass.meta()).setName("description").setContent("Material node example for X3D-Edit MaterialEditor, which uses Scene Access Interface (SAI) externally to select/deselect different shapes, axes and lights while modifying Material values."))
    .addMeta((new autoclass.meta()).setName("creator").setContent("Don Brutzman"))
    .addMeta((new autoclass.meta()).setName("created").setContent("17 December 2006"))
    .addMeta((new autoclass.meta()).setName("modified").setContent("20 October 2019"))
    .addMeta((new autoclass.meta()).setName("subject").setContent("X3D-Edit example"))
    .addMeta((new autoclass.meta()).setName("identifier").setContent("https://www.web3d.org/x3d/content/examples/Basic/development/MaterialExample.x3d"))
    .addMeta((new autoclass.meta()).setName("reference").setContent("src/www.web3d.org/x3d/tools/X3dEdit3.3/X3D/src/org/web3d/x3d/templates/MaterialExample.x3d"))
    .addMeta((new autoclass.meta()).setName("generator").setContent("X3D-Edit 3.3, https://savage.nps.edu/X3D-Edit"))
    .addMeta((new autoclass.meta()).setName("license").setContent("../license.html")))
  .setScene((new autoclass.Scene())
    .addChild((new autoclass.WorldInfo()).setTitle("MaterialExample.x3d"))
    .addChild((new autoclass.Viewpoint("ViewpointNode")).setDescription("default").setPosition(0.0,0.0,4.0))
    .addChild((new autoclass.NavigationInfo()).setType(new autoclass.MFString("\"EXAMINE\"")).setHeadlight(false))
    .addChild((new autoclass.Background("BackgroundNode")))
    .addChild((new autoclass.DirectionalLight("DirectionalLightNode")).setDirection(-0.707,0.0,-0.707))
    .addChild((new autoclass.Shape("ShapeNode"))
      .setGeometry((new autoclass.Sphere()))
      .setAppearance((new autoclass.Appearance())
        .setMaterial((new autoclass.Material("MaterialNode")))))
    .addChild((new autoclass.Switch("AxesSwitch")).setWhichChoice(0)
      .addChild((new autoclass.Group())
        .addChild((new autoclass.Group("xAxis"))
          .addChild((new autoclass.Shape())
            .setGeometry((new autoclass.IndexedLineSet()).setCoordIndex(java.newArray("int", [0,1]))
              .setCoord((new autoclass.Coordinate()).setPoint(new autoclass.MFVec3f(java.newArray("float", [0.0,0.0,0.0,1.25,0.0,0.0])))))
            .setAppearance((new autoclass.Appearance("AppearanceRed"))
              .setMaterial((new autoclass.Material()).setDiffuseColor(0.0,0.0,0.0).setEmissiveColor(0.8,0.0,0.0))))
          .addChild((new autoclass.Transform()).setTranslation(1.4,0.0,0.0)
            .addChild((new autoclass.Billboard())
              .addChild((new autoclass.Shape())
                .setGeometry((new autoclass.Text()).setString(new autoclass.MFString("\"x\""))
                  .setFontStyle((new autoclass.FontStyle("LabelStyle")).setJustify(new autoclass.MFString("\"MIDDLE\" \"MIDDLE\"")).setSize(0.3)))
                .setAppearance((new autoclass.Appearance()).setUSE("AppearanceRed"))))))
        .addChild((new autoclass.Group("yAxis"))
          .addChild((new autoclass.Shape())
            .setGeometry((new autoclass.IndexedLineSet()).setCoordIndex(java.newArray("int", [0,1]))
              .setCoord((new autoclass.Coordinate()).setPoint(new autoclass.MFVec3f(java.newArray("float", [0.0,0.0,0.0,0.0,1.25,0.0])))))
            .setAppearance((new autoclass.Appearance("AppearanceGreen"))
              .setMaterial((new autoclass.Material()).setDiffuseColor(0.0,0.0,0.0).setEmissiveColor(0.0,0.8,0.0))))
          .addChild((new autoclass.Transform()).setTranslation(0.0,1.4,0.0)
            .addChild((new autoclass.Billboard()).setAxisOfRotation(0.0,0.0,0.0)
              .addChild((new autoclass.Shape())
                .setGeometry((new autoclass.Text()).setString(new autoclass.MFString("\"y\""))
                  .setFontStyle((new autoclass.FontStyle()).setUSE("LabelStyle")))
                .setAppearance((new autoclass.Appearance()).setUSE("AppearanceGreen"))))))
        .addChild((new autoclass.Group("zAxis"))
          .addChild((new autoclass.Shape())
            .setGeometry((new autoclass.IndexedLineSet()).setCoordIndex(java.newArray("int", [0,1]))
              .setCoord((new autoclass.Coordinate()).setPoint(new autoclass.MFVec3f(java.newArray("float", [0.0,0.0,0.0,0.0,0.0,1.25])))))
            .setAppearance((new autoclass.Appearance("AppearanceBlue"))
              .setMaterial((new autoclass.Material()).setDiffuseColor(0.0,0.0,0.0).setEmissiveColor(0.0,0.0,0.8))))
          .addChild((new autoclass.Transform()).setTranslation(0.0,0.0,1.4)
            .addChild((new autoclass.Billboard()).setAxisOfRotation(0.0,0.0,0.0)
              .addChild((new autoclass.Shape())
                .setGeometry((new autoclass.Text()).setString(new autoclass.MFString("\"z\""))
                  .setFontStyle((new autoclass.FontStyle()).setUSE("LabelStyle")))
                .setAppearance((new autoclass.Appearance()).setUSE("AppearanceBlue"))))))))
    .addChild((new autoclass.Switch("LightDirectionVectorSwitch")).setWhichChoice(0)
      .addChild((new autoclass.Transform()).setScale(1.25,1.25,1.25)
        .addChild((new autoclass.Shape())
          .setGeometry((new autoclass.IndexedLineSet()).setCoordIndex(java.newArray("int", [0,1]))
            .addComments(" negate direction-vector values so that line approaches from light direction towards origin ")
            .setCoord((new autoclass.Coordinate("LightDirectionCoordinateNode")).setPoint(new autoclass.MFVec3f(java.newArray("float", [0.707,0.0,0.707,0.0,0.0,0.0])))))
          .setAppearance((new autoclass.Appearance())
            .setMaterial((new autoclass.Material("LightDirectionLineMaterial")).setDiffuseColor(0.0,0.0,0.0).setEmissiveColor(1.0,1.0,1.0)))))));
  },
  // end of initialize() method


  /** Provide a shallow copy of the X3D model.
   * @return MaterialExample model
   */
  getX3dModel : function()
  {	  
	  return this.x3dModel;
  },
  
  /** Indicate X3DJSAIL validation results for this X3D model.
   * @return validation results plus exception information, if any
   */
  validateSelf : function()
  {
	var validationResult = "";
	var  exceptionResult = "";
	try
	{
		this.initialize();
		
		if ((this.getX3dModel() == null) || (this.getX3dModel().getHead() == null))
		{
			validationResult = "empty scene, nothing to validate. " + this.x3dModel.validate();
			return validationResult;
		}
		validationResult += this.x3dModel.validate(); // walk entire tree to validate correctness
	}
	catch (e)
	{
		exceptionResult = e; // report exception failures, if any
	}
	if  (exceptionResult === "" && validationResult === "")
	     return "success";
	else
	{
		var returnMessage = "";
		if  (exceptionResult !== "" && validationResult !== "")
			returnMessage += "\n*** ";
		returnMessage += exceptionResult;
		if  (exceptionResult === "" && validationResult !== "")
			returnMessage = "\n" + returnMessage; // skip line before meta tags, etc.
		returnMessage += validationResult;
		return returnMessage;
	}
  },
    /** Default main() method provided for test purposes.
     * @param argv input parameters
     */
    main : function (argv)
    {
		var testObject = new MaterialExample();
		console.log ("MaterialExample execution self-validation test results: " + testObject.validateSelf());
	}
}
new MaterialExample().main();
process.exit(0);